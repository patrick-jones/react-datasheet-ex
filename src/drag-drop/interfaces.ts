import RDS from 'react-datasheet';
import {ConnectDragPreview, ConnectDragSource, ConnectDropTarget} from 'react-dnd';
import {ColumnHeader, HeaderRendererProps, RowRendererProps} from '../interfaces';

/**
 * Properties that must be provided to a drag and drop enabled row
 */
export interface DragDropRowProps<T extends RDS.Cell<T, V>, V = string> {
  /** The current row index */
  row: number;
  /** Callback for when a row is dropped on another */
  readonly onRowDropped: (source: number, target: number) => void;
}

/**
 * Properties that must be provided to a drag and drop enabled column header
 */
export interface DragDropHeaderProps {
  /** The current column index */
  col: number;
  /** Callback for one header is dropped on another */
  readonly onHeaderDropped: (source: number, target: number) => void;
}

/**
 * Drag source properties that are injected into a drag and drop provider
 */
export interface InjectedDragSourceProps {
  readonly connectDragSource: ConnectDragSource;
  readonly connectDragPreview: ConnectDragPreview;
}

/**
 * Drop target properties that are injected into a drag and drop provider
 */
export interface InjectedDropTargetProps {
  readonly connectDropTarget: ConnectDropTarget;
  readonly isDropOver: boolean;
}

/**
 * Combined drag source and drop target properties that are injected
 * into a drag and drop provider
 */
export interface InjectedDragDropProps extends InjectedDragSourceProps, InjectedDropTargetProps {}

/**
 * Extends `ReactDatasheet.RowRendererProps` to include
 * drag and drop properties
 */
export interface DragDropRowRendererProps<CellType, V = string>
  extends RowRendererProps<CellType, V>,
    DragDropRowProps<CellType, V> {}

/**
 * Extends `HeaderRendererProps` to include drag and drop properties.
 */
export interface DragDropHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends HeaderRendererProps<T>,
    DragDropHeaderProps {}
