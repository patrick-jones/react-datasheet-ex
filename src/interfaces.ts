import React, {CSSProperties} from 'react';
import RDS from 'react-datasheet';

export interface RowLocation {
  row: number;
}

export interface ColumnLocation {
  col: number;
}

export interface ColumnHeader {
  id: string;
  title: string;
  width?: number;
}

export interface HeaderRendererProps<T extends ColumnHeader = ColumnHeader> extends ColumnLocation {
  header: T;
}

export type HeaderRenderer<T extends ColumnHeader = ColumnHeader,
  R extends HeaderRendererProps<T> = HeaderRendererProps<T>> = React.ComponentType<R>;

export type HeaderContentRenderer<T extends ColumnHeader = ColumnHeader>
  = React.ComponentType<HeaderRendererProps<T>>;

export interface SheetRendererProps<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  >
  extends RDS.SheetRendererProps<T, V> {
  /** React style object that should be applied to the enclosing table */
  style?: CSSProperties | undefined;
  /**
   * An array of `ColumnHeader`, which are rendered into `th` elements
   * @default []
   */
  headers: H[];
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderRenderer
   */
  headerRenderer: HeaderRenderer;
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderContentRenderer
   */
  headerContentRenderer: HeaderContentRenderer;
}


export type ActionContentRenderer<T extends RowLocation = RowLocation>
  = React.ComponentType<T>;

export interface RowRendererProps<
    T extends RDS.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends RDS.RowRendererProps<T, V> {

  actionContentRenderer?: ActionContentRenderer<AP>;
}
