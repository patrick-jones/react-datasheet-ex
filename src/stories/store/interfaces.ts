import RDS from 'react-datasheet';
import {ColumnHeader, Overflow} from '../../interfaces';

export type ExampleCellValue = string | number | null;
export interface ExampleCellType extends RDS.Cell<ExampleCellType> {
  value: ExampleCellValue;
}

export interface ExampleModel {
  id: string;
  headers: ExampleHeader[];
  rows: ExampleRow[];
  selected: RDS.Selection | null;
  overflow: Overflow;
}

export interface ExampleHeader extends ColumnHeader {
  hidden?: boolean;
}

export interface ExampleRow {
  id: string;
  selected?: boolean;
  data: ExampleCellType[];
}

