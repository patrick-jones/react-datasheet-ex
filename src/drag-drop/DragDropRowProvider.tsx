import {Component, ReactNode} from 'react';

import { RowDragSource, RowDropTarget } from './row-dnd';
import { InjectedDragDropProps, DragDropRowProps } from './interfaces';


/**
 * Required properties for `DragDropRowProvider`
 */
export interface DragDropRowProviderProps<CellType, V = string>
  extends DragDropRowProps<CellType, V>, InjectedDragDropProps {
  children: (props: InjectedDragDropProps) => ReactNode;
}


class DragDropRow<CellType, V = string>
  extends Component<DragDropRowProviderProps<CellType, V>> {

  render(): ReactNode {
    const {children, ...rest} = this.props;
    return children(rest);
  }
}

/**
 * Provides all drag-and-drop properties to a render function assigned to the
 * `children` property.
 */
const DragDropRowProvider = RowDropTarget(RowDragSource(DragDropRow));

export default DragDropRowProvider;
