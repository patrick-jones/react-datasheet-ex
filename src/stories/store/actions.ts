import RDS from 'react-datasheet';
import {ExampleCellType} from './interfaces';
import {Overflow} from '../../interfaces';

/**
 * Some quick redux-like interfaces
 */

/**
 * A redux-like action
 */
export type Action<P = any> = {
  readonly type: string;
  readonly payload: P;
};

/**
 * Abstract type for a function that creates an action
 */
export type ActionCreator = (...args: any[]) => Action;

/**
 * Abstract type for a mapping of keys are ActionCreator functions
 */
export type ActionsCreators = {
  [key: string]: ActionCreator;
};

/**
 * Abstract type for the Action objects returned by an ActionCreators mapping
 */
export type Actions<A extends ActionsCreators> = ReturnType<A[keyof A]>;

/**
 * Type for actions once they are run through the dispatch function
 */
export type DispatchedActionCreator<AC extends ActionCreator> = (...args: any[]) => void;

/**
 * Converts `actions` object from functions that return an Action to
 * functions with the same arguments that return void.
 */
export type DispatchedActions<TAC> = {
  [K in keyof TAC]: TAC[K] extends ((...args: infer A) => Action) ? (...args: A) => void : undefined
};

const createAction = <T extends string, P>(type: T, payload: P) => ({type, payload});

/**
 * Our specific actions
 */
export enum ActionTypes {
  CELLS_CHANGED = 'rds-ex__cells-changed',
  SELECTION_CHANGED = 'rds-ex__selection-changed',
  OVERFLOW_CHANGED = 'rds-ex__overflow-changed',
  HEADER_DROPPED = 'rds-ex__header-dropped',
  ROW_DROPPED = 'rds-ex__row-dropped',
  ROW_SELECTION_CHANGED = 'rds-ex__row-selection-changed',
  COLUMN_HIDDEN_CHANGED = 'rds-ex__column-hidden-changed',
  HEADER_RESIZING = 'rds-ex__header-resizing',
  HEADER_RESIZE_END = 'rds-ex__header-resize-end',
}

export const SheetActionCreators = {
  cellsChanged:
    (changes: RDS.CellsChangedArgs<ExampleCellType>, additions?: RDS.CellAdditionsArgs<string>) =>
      createAction(ActionTypes.CELLS_CHANGED, {changes, additions}),
  selectionChanged: (selected: RDS.Selection) =>
    createAction(ActionTypes.SELECTION_CHANGED, {selected}),
  overflowChanged: (overflow: Overflow) =>
    createAction(ActionTypes.OVERFLOW_CHANGED, {overflow}),
  headerDropped: (source: number, target: number) =>
    createAction(ActionTypes.HEADER_DROPPED, {source, target}),
  rowDropped: (source: number, target: number) =>
    createAction(ActionTypes.ROW_DROPPED, {source, target}),
  rowSelectionChanged: (row: number, selected: boolean) =>
    createAction(ActionTypes.ROW_SELECTION_CHANGED, {row, selected}),
  columnHiddenChanged: (id: string, hidden: boolean) =>
    createAction(ActionTypes.COLUMN_HIDDEN_CHANGED, {id, hidden}),
  headerResizing: (source: number, requestedSize: number) =>
    createAction(ActionTypes.HEADER_RESIZING, {source, requestedSize}),
  headerResizeEnd: (source: number, requestedSize: number) =>
    createAction(ActionTypes.HEADER_RESIZE_END, {source, requestedSize}),
};

export type SheetActionCreators = typeof SheetActionCreators;
