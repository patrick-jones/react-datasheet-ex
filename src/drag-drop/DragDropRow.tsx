import React, {Component} from 'react';

import {ClassNames} from '../';
import DefaultAction from '../DefaultAction';

import {InjectedDragDropProps, DragDropRowRendererProps} from './interfaces';
import DragDropRowProvider from './DragDropRowProvider';
import {ActionRendererProps} from '../interfaces';
import RDS from 'react-datasheet';


const cellClasses = [
  ClassNames.READ_ONLY_CELL,
  ClassNames.ACTION_CELL,
  ClassNames.ROW_DRAG_SOURCE,
].join(' ');


const rowClass = ClassNames.ROW;
const dtClass = [ClassNames.ROW, ClassNames.ROW_DROP_TARGET].join(' ');


export default class DragDropRow<
  CellType extends RDS.Cell<CellType, V>, V = string,
  ARP extends ActionRendererProps<CellType, V> = ActionRendererProps<CellType, V>
>
  extends Component<DragDropRowRendererProps<CellType, V, ARP>> {

  static readonly defaultProps = {
    actionRenderer: DefaultAction,
  };

  renderRow = (dndProps: InjectedDragDropProps) => {
    const {connectDragSource, connectDragPreview, connectDropTarget, isDropOver} = dndProps;
    const {id, row, children, actionRenderer: ActionRenderer, ...rest} = this.props;

    return connectDropTarget(connectDragPreview(
      <tr key={`$row-${id || row}`} className={isDropOver ? dtClass : rowClass}>
        {connectDragSource(
          <td key='$$actionCell' className={cellClasses}>
            <ActionRenderer id={id} row={row} {...rest} />
          </td>
        )}
        {children}
      </tr>
    ));
  };

  render() {
    const {row, onRowDropped} = this.props;

    return (
      <DragDropRowProvider row={row} onRowDropped={onRowDropped}>
        {this.renderRow}
      </DragDropRowProvider>
    );
  }
}
