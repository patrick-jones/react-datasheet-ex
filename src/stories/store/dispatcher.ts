import {Store} from '@dump247/storybook-state';
import {ExampleModel} from './interfaces';
import {Actions, ActionsCreators, DispatchedActionCreator, DispatchedActions} from './actions';
import reducer from './reducer';
import {action} from '@storybook/addon-actions';

function dispatch<AC extends ActionsCreators>(store: Store<ExampleModel>, a: Actions<AC>) {
  action(a.type)(a.payload);
  const nextState = reducer(store.state, a);
  if (nextState && nextState !== store.state) {
    store.set(nextState);
  }
}

export default function bindDispatch<AC extends ActionsCreators>(
  store: Store<ExampleModel>,
  actionCreators: AC
): DispatchedActions<AC> {
  const bound = {} as any;
  Object.keys(actionCreators).forEach(key => {
    const orig = actionCreators[key];
    const ds: DispatchedActionCreator<typeof orig> = (...args) =>
      dispatch(store, actionCreators[key](...args));
    bound[key] = ds;
  });
  return bound as DispatchedActions<AC>;
}
