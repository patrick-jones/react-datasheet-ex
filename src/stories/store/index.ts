import {model} from './model';
import {ExampleModel} from './interfaces';
import {Actions, SheetActionCreators} from './actions';
import reducer from './reducer';

export {model};

export function preprocess(model: ExampleModel, ...actions: Actions<SheetActionCreators>[]) {
  return actions.reduce(
    (reply, action) => {
      const nextState = reducer(reply, action);
      return Object.assign({}, reply, nextState);
    },
    model
  );
}
