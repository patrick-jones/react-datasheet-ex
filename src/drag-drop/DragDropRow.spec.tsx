import React from 'react';
import 'jest';
import {mount} from 'enzyme';

import {DragDropContext} from 'react-dnd';
import Backend, {TestBackend} from 'react-dnd-test-backend';
import toJson from 'enzyme-to-json';

import DragDropRow from './DragDropRow';
import DragDropRowProvider from './DragDropRowProvider';
import {ExampleCellType} from '../stories/store/interfaces';

const rows: ExampleCellType[][] = [
  [{value: 'Row 0'}],
  [{value: 'Row 1'}],
];

function testableComponent(spyA: jest.Mock, spyB: jest.Mock) {
  return DragDropContext(Backend)(
    class TestContainer extends React.Component {
      render() {
        return (
          <table>
            <tbody>
            <DragDropRow row={0} cells={rows[0]} onRowDropped={spyA}>
              <td>Row 0</td>
            </DragDropRow>
            <DragDropRow row={1} cells={rows[1]} onRowDropped={spyB}>
              <td>Row 1</td>
            </DragDropRow>
            </tbody>
          </table>
        );
      }
    }
  );
}


describe('DragDropRowRenderer', () => {
  it('matches snapshot', () => {
    const spyA = jest.fn();
    const spyB = jest.fn();
    const Container = testableComponent(spyA, spyB);

    const wrapper = mount(<Container />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Calls onHeaderDropped', () => {
    const spyA = jest.fn();
    const spyB = jest.fn();
    const Container = testableComponent(spyA, spyB);

    const wrapper = mount(<Container />);
    // console.log(wrapper.debug());
    // @ts-ignore
    const backend: TestBackend = wrapper.instance().getManager().getBackend();

    const hp = wrapper.find(DragDropRowProvider);

    // @ts-ignore
    const dragSource = hp.at(1).childAt(0).instance().getHandlerId();
    // @ts-ignore
    const dropTarget = hp.at(0).instance().getHandlerId();

    backend.simulateBeginDrag([dragSource]);
    backend.simulateHover([dropTarget]);
    backend.simulateDrop();
    backend.simulateEndDrag();

    expect(spyA.mock.calls.length).toBe(1);
    expect(spyB.mock.calls.length).toBe(0);

    expect(spyA.mock.calls[0][0]).toBe(1);
    expect(spyA.mock.calls[0][1]).toBe(0);
  });
});
