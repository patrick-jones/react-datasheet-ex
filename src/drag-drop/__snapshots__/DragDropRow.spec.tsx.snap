// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragDropRowRenderer matches snapshot 1`] = `
<DragDropContext(TestContainer)>
  <TestContainer>
    <table>
      <tbody>
        <DragDropRow
          actionRenderer={[Function]}
          cells={
            Array [
              Object {
                "value": "Row 0",
              },
            ]
          }
          onRowDropped={[MockFunction]}
          row={0}
        >
          <DropTarget(DragSource(DragDropRow))
            onRowDropped={[MockFunction]}
            row={0}
          >
            <DragSource(DragDropRow)
              connectDropTarget={[Function]}
              isDropOver={false}
              onRowDropped={[MockFunction]}
              row={0}
            >
              <DragDropRow
                connectDragPreview={[Function]}
                connectDragSource={[Function]}
                connectDropTarget={[Function]}
                isDropOver={false}
                onRowDropped={[MockFunction]}
                row={0}
              >
                <tr
                  className="rdx-row"
                  key="$row-0"
                >
                  <td
                    className="cell read-only rdx-action-cell rdx-row-drag-source"
                    key="$$actionCell"
                  >
                    <DefaultAction
                      cells={
                        Array [
                          Object {
                            "value": "Row 0",
                          },
                        ]
                      }
                      onRowDropped={[MockFunction]}
                      row={0}
                    />
                  </td>
                  <td>
                    Row 0
                  </td>
                </tr>
              </DragDropRow>
            </DragSource(DragDropRow)>
          </DropTarget(DragSource(DragDropRow))>
        </DragDropRow>
        <DragDropRow
          actionRenderer={[Function]}
          cells={
            Array [
              Object {
                "value": "Row 1",
              },
            ]
          }
          onRowDropped={[MockFunction]}
          row={1}
        >
          <DropTarget(DragSource(DragDropRow))
            onRowDropped={[MockFunction]}
            row={1}
          >
            <DragSource(DragDropRow)
              connectDropTarget={[Function]}
              isDropOver={false}
              onRowDropped={[MockFunction]}
              row={1}
            >
              <DragDropRow
                connectDragPreview={[Function]}
                connectDragSource={[Function]}
                connectDropTarget={[Function]}
                isDropOver={false}
                onRowDropped={[MockFunction]}
                row={1}
              >
                <tr
                  className="rdx-row"
                  key="$row-1"
                >
                  <td
                    className="cell read-only rdx-action-cell rdx-row-drag-source"
                    key="$$actionCell"
                  >
                    <DefaultAction
                      cells={
                        Array [
                          Object {
                            "value": "Row 1",
                          },
                        ]
                      }
                      onRowDropped={[MockFunction]}
                      row={1}
                    />
                  </td>
                  <td>
                    Row 1
                  </td>
                </tr>
              </DragDropRow>
            </DragSource(DragDropRow)>
          </DropTarget(DragSource(DragDropRow))>
        </DragDropRow>
      </tbody>
    </table>
  </TestContainer>
</DragDropContext(TestContainer)>
`;
